Merge frontend and backend into a monorepo

kalkins opened the pull request:
The backend has been added as a subtree, meaning that git retains its history.

This does however create some weird history conflicts when following files through renames. If a file had the same name in frontend and backend (like README.md), git merges their history when showing the log. This does not impact the files or the repo log, only the log retrieved for that one file.

kalkins commented:
@FredrikAugust @KaizerTilt @NicolajNiels1 @joachifm 
I've set up docker and circle-ci. The website works, but the tests fail. Locally there are some recent changes in the backend that aren't reflected in the frontend. @FredrikAugust will fix that soon.

However, the frontend tests don't run at all in circle-ci because `react-scripts` isn't installed. Does anyone have any idea why this works locally but not on circle-ci?

kalkins commented:
@joachifm You seem to be experienced with docker. Do you see anything that can be improved in the docker and docker-compose files?

joachifm commented:
@kalkins one other thing occurred to me: though it breaks running migrations inside the container, running the apps readonly in production would be prudent. I believe this can be achieved fairly cleanly by supplying a `prod` override compose file, that sets `read_only: true` on the services & mounts `/code` read-only.

FredrikAugust commented:
`react-scripts` has to be installed globally (or be in PATH) for it to
work. It should work out of the box after running yarn, but you could run
the local binary instead - node_modules/.bin/react-scripts. That could
work. It did work just fine in CircleCI before though, so I think something
might be wrong with the setup of that.

El sáb., 10 ago. 2019 16:04, Joachim F. <notifications@github.com> escribió:

> @kalkins <https://github.com/kalkins> one other thing occurred to me:
> though it breaks running migrations inside the container, running the apps
> readonly in production would be prudent. I believe this can be achieved
> fairly cleanly by supplying a prod override compose file, that sets read_only:
> true on the services & mounts /code read-only.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2GYFVUMTT7B7DIGC2Y3QD3DHLA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4AOEHY#issuecomment-520151583>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2G4PMCPIWYZ2G3QX42TQD3DHLANCNFSM4IKNEQPQ>
> .
>

kalkins commented on backend/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+RUN ./manage.py collectstatic --no-input
+
+EXPOSE 8000
```
I was following a guide that included this. I'll remove it if it's not needed

kalkins commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM node:latest
```
I took the same tag as the current circle-ci setup is using. @FredrikAugust Do you have any suggestions here?

FredrikAugust replied:
I would argue that always using the latest node version is better. It will most likely net us some performance improvements further down the line, and it should never break anything.

kalkins commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM node:latest
+
+#RUN useradd frontend -u 1000 -U -s /bin/false
```
I tried it, the same way it's used in the backend, but got an error. It's a low priority right now, but I'll take a look at it after fixing circle-ci

kalkins replied:
Opps, didn't read through your comment properly. Yes, I can try the node user

FredrikAugust commented:
There might be an image that ships with yarn, which would be better as you
need it. After that, just run `yarn` before doing anything with the
frontend - this installs the deps.

El sáb., 10 ago. 2019 17:55, Sindre Stephansen <notifications@github.com>
escribió:

> *@kalkins* commented on this pull request.
> ------------------------------
>
> In frontend/Dockerfile
> <https://github.com/Kodeworks/liquidator-frontend/pull/128#discussion_r312707359>
> :
>
> > @@ -0,0 +1,18 @@
> +FROM node:latest
> +
> +#RUN useradd frontend -u 1000 -U -s /bin/false
>
> Opps, didn't read through your comment properly. Yes, I can try the node
> user
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G4X273ZUAZYRUR6XDLQD3QGJA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBGDOZQ#discussion_r312707359>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2G25KAGZKM3EXUNSRNDQD3QGJANCNFSM4IKNEQPQ>
> .
>

kalkins commented:
There was a seperate yarn image, but it's deprecated as yarn is provided with the standard node image

FredrikAugust commented:
Alright, great.

El sáb., 10 ago. 2019 17:59, Sindre Stephansen <notifications@github.com>
escribió:

> There was a seperate yarn image, but it's deprecated as yarn is provided
> with the standard node image
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G6LX2MVUQ2DEEAVE63QD3QVVA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4AQEXY#issuecomment-520159839>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2GZM2NBT2N6PFAD74TLQD3QVVANCNFSM4IKNEQPQ>
> .
>

kalkins commented:
Does yarn only need package.json and yarn.lock when getting dependencies? It would improve caching if we could copy over those files, install dependencies, and then copy over the rest of the project

FredrikAugust commented:
Correct. It only needs those two.

El sáb., 10 ago. 2019 18:01, Sindre Stephansen <notifications@github.com>
escribió:

> Does yarn only need package.json and yarn.lock when getting dependencies?
> It would improve caching if we could copy over those files, install
> dependencies, and then copy over the rest of the project
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G74X7RINUUMFD7ZZE3QD3Q4PA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4AQFZI#issuecomment-520159973>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2GZIN7WSKZNZ37FYSELQD3Q4PANCNFSM4IKNEQPQ>
> .
>

kalkins commented:
> Does this use yarn start in prod? Because if so, it should be changed to run yarn buld, and then serve the files staticly

I used `yarn start` because it was easier while setting up docker, and making sure they worked together. We have to do something about that before moving to prod. This is also related to something Joachim said earlier:

> @kalkins one other thing occurred to me: though it breaks running migrations inside the container, running the apps readonly in production would be prudent. I believe this can be achieved fairly cleanly by supplying a `prod` override compose file, that sets `read_only: true` on the services & mounts `/code` read-only.

I was thinking of doing something similar to the run commands, so we get hot reload while developing, and production builds in production. @joachifm Could we do this with environment variables instead of separate compose files? So we have `read_only: ${PROD}` or something like that? For the run commands we could either have a one-line bash command or a separate script that starts the dev server, or does a prod build

FredrikAugust commented:
Sounds good to me, as long as we don't run yarn start in prod we should be
all good!

El sáb., 10 ago. 2019 18:23, Sindre Stephansen <notifications@github.com>
escribió:

> Does this use yarn start in prod? Because if so, it should be changed to
> run yarn buld, and then serve the files staticly
>
> I used yarn start because it was easier while setting up docker, and
> making sure they worked together. We have to do something about that before
> moving to prod. This is also related to something Joachim said earlier:
>
> @kalkins <https://github.com/kalkins> one other thing occurred to me:
> though it breaks running migrations inside the container, running the apps
> readonly in production would be prudent. I believe this can be achieved
> fairly cleanly by supplying a prod override compose file, that sets read_only:
> true on the services & mounts /code read-only.
>
> I was thinking of doing something similar to the run commands, so we get
> hot reload while developing, and production builds in production.
> @joachifm <https://github.com/joachifm> Could we do this with environment
> variables instead of separate compose files? So we have read_only: ${PROD}
> or something like that? For the run commands we could either have a
> one-line bash command or a separate script that starts the dev server, or
> does a prod build
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G7L3NRKQLXM4VU7W4LQD3TR5A5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4AQUMQ#issuecomment-520161842>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2G5BZHZIHHGG6M4HUTLQD3TR5ANCNFSM4IKNEQPQ>
> .
>

kalkins commented:
> `react-scripts` has to be installed globally (or be in PATH) for it to work. It should work out of the box after running yarn, but you could run the local binary instead - node_modules/.bin/react-scripts. That could work. It did work just fine in CircleCI before though, so I think something might be wrong with the setup of that.

The weird thing is, the error in circle-ci is `/bin/sh: react-scripts: not found`. When I try to run `react-scripts` as part of the Dockerfile locally, I get the same error. However, if I try to do `docker-compose run frontend react-scripts`, I get this error:

```
internal/modules/cjs/loader.js:716
    throw err;
    ^

Error: Cannot find module '/frontend/code/react-scripts'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:713:15)
    at Function.Module._load (internal/modules/cjs/loader.js:618:27)
    at Function.Module.runMain (internal/modules/cjs/loader.js:931:10)
    at internal/main/run_main_module.js:17:11 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

When I run `docker-compose run frontend yarn test` it says it runs `react-scripts test --testPathIgnorePatterns "mitochondria"`, which works fine locally, but not in circle-ci.

FredrikAugust commented:
There probably is a problem with path resolving or something along those
lines.

I do not know how this happens, but the way it should ideally be run is
like this:

1. CD into dir
2. Yarn
3. Yarn start

This way the working directory is the root of the frontend.

El sáb., 10 ago. 2019 18:48, Sindre Stephansen <notifications@github.com>
escribió:

> react-scripts has to be installed globally (or be in PATH) for it to
> work. It should work out of the box after running yarn, but you could run
> the local binary instead - node_modules/.bin/react-scripts. That could
> work. It did work just fine in CircleCI before though, so I think something
> might be wrong with the setup of that.
>
> The weird thing is, the error in circle-ci is /bin/sh: react-scripts: not
> found. When I try to run react-scripts as part of the Dockerfile locally,
> I get the same error. However, if I try to do docker-compose run frontend
> react-scripts, I get this error:
>
> internal/modules/cjs/loader.js:716
>     throw err;
>     ^
>
> Error: Cannot find module '/frontend/code/react-scripts'
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:713:15)
>     at Function.Module._load (internal/modules/cjs/loader.js:618:27)
>     at Function.Module.runMain (internal/modules/cjs/loader.js:931:10)
>     at internal/main/run_main_module.js:17:11 {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: []
> }
>
> When I run docker-compose run frontend yarn test it says it runs react-scripts
> test --testPathIgnorePatterns "mitochondria", which works fine locally,
> but not in circle-ci.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G2LA53HD7WOX6YNZQLQD3WNTA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ARDDA#issuecomment-520163724>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2G6D7SKLETTXHGDPVOTQD3WNTANCNFSM4IKNEQPQ>
> .
>

kalkins commented:
Thats what we do. Is there a way to install/set up the path to react-scripts manually so we are sure it will be available?


On Sat, Aug 10, 2019, at 18:54, Fredrik August Madsen-Malmo wrote:
> There probably is a problem with path resolving or something along those
>  lines.
> 
>  I do not know how this happens, but the way it should ideally be run is
>  like this:
> 
>  1. CD into dir
>  2. Yarn
>  3. Yarn start
> 
>  This way the working directory is the root of the frontend.
> 
>  El sáb., 10 ago. 2019 18:48, Sindre Stephansen <notifications@github.com>
>  escribió:
> 
>  > react-scripts has to be installed globally (or be in PATH) for it to
>  > work. It should work out of the box after running yarn, but you could run
>  > the local binary instead - node_modules/.bin/react-scripts. That could
>  > work. It did work just fine in CircleCI before though, so I think something
>  > might be wrong with the setup of that.
>  >
>  > The weird thing is, the error in circle-ci is /bin/sh: react-scripts: not
>  > found. When I try to run react-scripts as part of the Dockerfile locally,
>  > I get the same error. However, if I try to do docker-compose run frontend
>  > react-scripts, I get this error:
>  >
>  > internal/modules/cjs/loader.js:716
>  > throw err;
>  > ^
>  >
>  > Error: Cannot find module '/frontend/code/react-scripts'
>  > at Function.Module._resolveFilename (internal/modules/cjs/loader.js:713:15)
>  > at Function.Module._load (internal/modules/cjs/loader.js:618:27)
>  > at Function.Module.runMain (internal/modules/cjs/loader.js:931:10)
>  > at internal/main/run_main_module.js:17:11 {
>  > code: 'MODULE_NOT_FOUND',
>  > requireStack: []
>  > }
>  >
>  > When I run docker-compose run frontend yarn test it says it runs react-scripts
>  > test --testPathIgnorePatterns "mitochondria", which works fine locally,
>  > but not in circle-ci.
>  >
>  > —
>  > You are receiving this because you were mentioned.
>  > Reply to this email directly, view it on GitHub
>  > <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G2LA53HD7WOX6YNZQLQD3WNTA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ARDDA#issuecomment-520163724>,
>  > or mute the thread
>  > <https://github.com/notifications/unsubscribe-auth/AB2V2G6D7SKLETTXHGDPVOTQD3WNTANCNFSM4IKNEQPQ>
>  > .
>  >
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB4GLNEJ4UK7J7PPJ3P3K6TQD3XEDA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ARGDI#issuecomment-520164109>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4GLNB4N7BTQ3MLF7TSOYDQD3XEDANCNFSM4IKNEQPQ>.

FredrikAugust commented:
I replied to that earlier in the thread. On mobile right now so finding
that is a bit tedious.

El sáb., 10 ago. 2019 18:57, Sindre Stephansen <notifications@github.com>
escribió:

> Thats what we do. Is there a way to install/set up the path to
> react-scripts manually so we are sure it will be available?
>
>
> On Sat, Aug 10, 2019, at 18:54, Fredrik August Madsen-Malmo wrote:
> > There probably is a problem with path resolving or something along those
> > lines.
> >
> > I do not know how this happens, but the way it should ideally be run is
> > like this:
> >
> > 1. CD into dir
> > 2. Yarn
> > 3. Yarn start
> >
> > This way the working directory is the root of the frontend.
> >
> > El sáb., 10 ago. 2019 18:48, Sindre Stephansen <notifications@github.com
> >
> > escribió:
> >
> > > react-scripts has to be installed globally (or be in PATH) for it to
> > > work. It should work out of the box after running yarn, but you could
> run
> > > the local binary instead - node_modules/.bin/react-scripts. That could
> > > work. It did work just fine in CircleCI before though, so I think
> something
> > > might be wrong with the setup of that.
> > >
> > > The weird thing is, the error in circle-ci is /bin/sh: react-scripts:
> not
> > > found. When I try to run react-scripts as part of the Dockerfile
> locally,
> > > I get the same error. However, if I try to do docker-compose run
> frontend
> > > react-scripts, I get this error:
> > >
> > > internal/modules/cjs/loader.js:716
> > > throw err;
> > > ^
> > >
> > > Error: Cannot find module '/frontend/code/react-scripts'
> > > at Function.Module._resolveFilename
> (internal/modules/cjs/loader.js:713:15)
> > > at Function.Module._load (internal/modules/cjs/loader.js:618:27)
> > > at Function.Module.runMain (internal/modules/cjs/loader.js:931:10)
> > > at internal/main/run_main_module.js:17:11 {
> > > code: 'MODULE_NOT_FOUND',
> > > requireStack: []
> > > }
> > >
> > > When I run docker-compose run frontend yarn test it says it runs
> react-scripts
> > > test --testPathIgnorePatterns "mitochondria", which works fine locally,
> > > but not in circle-ci.
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <
> https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G2LA53HD7WOX6YNZQLQD3WNTA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ARDDA#issuecomment-520163724
> >,
> > > or mute the thread
> > > <
> https://github.com/notifications/unsubscribe-auth/AB2V2G6D7SKLETTXHGDPVOTQD3WNTANCNFSM4IKNEQPQ
> >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub <
> https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB4GLNEJ4UK7J7PPJ3P3K6TQD3XEDA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ARGDI#issuecomment-520164109>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AB4GLNB4N7BTQ3MLF7TSOYDQD3XEDANCNFSM4IKNEQPQ
> >.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kodeworks/liquidator-frontend/pull/128?email_source=notifications&email_token=AB2V2G76J4ZDKSMZ4JJLSTLQD3XPJA5CNFSM4IKNEQP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ARH4A#issuecomment-520164336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2V2G25U3XMCIIJ4YHDTGLQD3XPJANCNFSM4IKNEQPQ>
> .
>

kalkins commented:
I've found and fixed the issue with `react-scripts` now. The problem was that the node modules were installed in the workdir, which was then overwritten by the volume mount. The reason it worked for me was that i had the modules installed from before. I fixed it by moving the node modules up to the parent folder.

kalkins commented:
The CI works now, there are only the three tests i mentioned previously that still fails.

I have set up the api tests so they run in a separate project, meaning that the containers and volumes don't interfere with the ones we use normally. The database is reset before each test run, so if a test fails the database can be inspected, but it won't affect future tests.

FredrikAugust commented:
Seems like `tslint` isn't available. @kalkins could you take a look at this?

kalkins commented:
@FredrikAugust I've fixed tslint, so all the checks pass. I'm not sure if pep8speaks works with this setup (and it isn't activated for this repo yet), but we can fix that later

kalkins commented:
@joachifm Do you have any thoughts about this: https://github.com/Kodeworks/liquidator-frontend/pull/128#issuecomment-520161842

joachifm commented:
@kalkins 

> Could we do this with environment variables instead of separate compose files? So we have read_only: ${PROD} or something like that? For the run commands we could either have a one-line bash command or a separate script that starts the dev server, or does a prod build

I'm confident this can be accomplished with environment variables as well.  But to elaborate a little bit on the case for a declarative approach with compose files:

The compose file approach only requires listing the keys that should be overridden.  Suppose you have `docker-compose.{prod,dev}.yml`, then a wrapper could do `docker-compose -f docker-compose.yml -f docker-compose.${ENV}.yml up -d`. The final `docker-compose.yml` should be the result of merging all the listed files.

I have a patch against the old backend repo that adds a `docker-compose.prod.yml` that contains only
```yaml
services:
  web:
    read_only: true
    volumes:
      - .:/code:ro
```

I think this could be simplified further by omitting the bind-mount of pwd into the container from the base composition and only doing that in the `dev` override. To my eyes this is fairly low-overhead.

I believe this would work for run instructions as well (just declare different commands in dev and prod, or change relevant envvars to use different ports or whatever).

If different build instructions are desired, it seems most efficient to me to use a layerered approach, where dev is derived from prod (or both derived from a common base, whichever is most convenient).

kalkins commented:
@joachifm I realized how it could be done shortly after pinging you, and I'm working on it now. 

Using environment variables would be complicated and hacky, and the fact that we can override the relevant parts of the compose file makes it a better option. Currently I'm trying to use `docker-compose.yml` as the base and `docker-compose.override.yml` as test/dev, as that is used by default. Then we have `docker-compose.prod.yml` which is used is production.

I agree with removing the bind mounts, and that is the approach I'm taking now. The files are copied during the docker build anyway.

Should we remove the final commands from the dockerfiles and have them in the compose file instead? That might make it more clear. Currently I'm just having the dev commands in the dockerfile and overriding them in the prod compose file.

kalkins commented:
@FredrikAugust The tests run fine, but the frontend fails to compile:
```
/frontend/code/src/store/reducers/company.ts
TypeScript error in /frontend/code/src/store/reducers/company.ts(75,30):
Argument of type 'ICompanyUser' is not assignable to parameter of type 'IAddCompanyUser'.
  Property 'email' is missing in type 'ICompanyUser' but required in type 'IAddCompanyUser'.  TS2345

    73 |   dispatch: React.Dispatch<ICreatedAction>
    74 | ) => {
  > 75 |   await API.addUserToCompany(user);
       |                              ^
    76 |   dispatch(addUserToCompany(user));
    77 | };
    78 |
```

joachifm commented:
@kalkins 

> Currently I'm just having the dev commands in the dockerfile and overriding them in the prod compose file.

If run instructions are kept in the dockerfile, I agree they should be for dev (or at least not expose the app to the open internet by default).

I think keeping everything in the compose files makes sense, even if for no other reason than consistency.

FredrikAugust commented:
And it now build successfully @kalkins.

FredrikAugust commented:
@kalkins can you merge this?

NicolajNiels1 commented on README.md:
```diff
@@ -5,7 +5,8 @@
 [**Milestones**](https://github.com/Kodeworks/liquidator-frontend/milestones) and [**Project Board**](https://github.com/Kodeworks/liquidator-frontend/projects/2)
 
-[**Link to backend repo**](https://github.com/kodeworks/liquidator-backend)
+[API spesification](https://app.swaggerhub.com/apis/kw-liquidator/Liquidator/1.0.0#/)
```
```suggestion
[API specification](https://app.swaggerhub.com/apis/kw-liquidator/Liquidator/1.0.0#/)
```

FredrikAugust commented:
@kalkins You just broke the tests with your last two commits. I'm getting a 502 Bad Gateway on all requests to the backend.

kalkins commented:
Eh, there's some weird django error. I reverted it so we still have the commit in the history
